### DISCUSSION TITLE
Report functions in let expressions that do not depend on local values

### DISCUSSION THREAD

**Discussion by jfmengels** [jfmengels/elm-review#93] [Rule ideas]
**What the rule should do:**

Report functions in let expressions that do not depend on local values

Prior work: `elm-analyse` has a rule to forbid functions in let expressions, regardless of how they are used.  https://stil4m.github.io/elm-analyse/#/messages/FunctionInLet
I don't like this rule and disabled it when I was using `elm-analyse` because functions could be using closures in an aim to make the code using it simpler.

**What problems does it solve:**

Making functions much smaller in a way that makes them much more readable IMO.

I don't know if defining functions inside let expressions adds a performance cost because the functions would be redefined at every call.

**Example of things the rule would report:**

```elm
kebabCase str =
  let
    toLowerCase s = String.toLowerCase s
  in
  magicAlgorithm str |> toLowerCase
```

**Example of things the rule would not report:**

```elm
toLowerCase s = String.toLowerCase s

kebabCase str =
  magicAlgorithm str |> toLowerCase

func n =
  let
    -- depends on `n`
    func m = m + n
  in
  func 1
```

**When (not) to enable this rule:**

Only enable if if you care about this kind of cleanup? :man_shrugging: 

**Questions I have**

- Should this be applied only on functions, or also on constants?

If so, in the following example `secondsInMinute` should be moved to the top-level.
```elm
toSeconds nbMinutes =
  let
    secondsInMinute = 60
  in
  nbMinutes * secondsInMinute
```

I'd say why not, but it might be too much? Note that point-free functions look like constants, especially without type annotations

- In what package would you see this being available?

**Comment by MartinSStewart** [comment]
> I don't like this rule and disabled it when I was using `elm-analyse` because functions could

Could what?
